cpp/inline/games/stochastic_nim/PerfectPlayer.inl

```diff
@@ -0,0 +1,58 @@
+#include <games/stochastic_nim/PerfectPlayer.hpp>
+
+namespace stochastic_nim {
+
+inline PerfectPlayer::ActionResponse PerfectPlayer::get_action_response(
+    const State& state, const ActionMask& valid_actions) {
+  util::release_assert(state.current_mode == kPlayerMode,
+                       "PerfectPlayer does not support chance mode");
+  return strategy_->get_optimal_action(state.stones_left);
+}
+
+PerfectStrategy::PerfectStrategy() {
+  for (int i = 0; i < stochastic_nim::kStartingStones + 1; i++) {
+    state_values_[i] = -1.0;
+    optimal_actions_[i] = -1;
+  }
+
+  state_values_[0] = 1.0;
+  state_values_[1] = 0.8;
+  state_values_[2] = 0.5;
+  state_values_[3] = 0.0;
+  optimal_actions_[1] = 1;
+  optimal_actions_[2] = 2;
+  optimal_actions_[3] = 3;
+  iterate();
+}
+
+inline int PerfectStrategy::get_optimal_action(int stones_left) const {
+  util::release_assert(
+      (stones_left <= stochastic_nim::kStartingStones) && (stones_left > 0),
+      "PerfectStrategy does not support more stones than starting stones or less or equal to 0");
+  return optimal_actions_[stones_left] - 1;
+}
+
+inline void PerfectStrategy::iterate() {
+  Eigen::Array<float, stochastic_nim::kChanceDistributionSize, 1> prob_array{
+      stochastic_nim::kChanceEventProbs};
+  auto reverse_probs = prob_array.reverse().eval();
+  for (int stones_left = 4; stones_left <= stochastic_nim::kStartingStones; stones_left++) {
+    optimal_actions_[stones_left] =
+        stochastic_nim::kMaxStonesToTake -
+        eigen_util::argmax(state_values_.segment(stones_left - stochastic_nim::kMaxStonesToTake,
+                                                 stochastic_nim::kMaxStonesToTake));
```

Comment:
Within tight-contexts like this function, it is perfectly acceptable, and even advisable, to introduce very-short variables to alias longer ones.

The general rule is: the wider the scope, the longer and more detailed your names should be. Conversely, the tighter the scope, the shorter they can be.

This line could read,

```
P[s] = m - argmax(Q.segment(s-m, m));
```

That is much clearer, simply by virtue of fitting in one line and being comprised of fewer characters.

To facilitate this, you would have above the loop,

```
auto& P = optimal_actions_;
constexpr int m = stochastic_nim::kMaxStonesToTake;
using namespace eigen_util;  // for argmax()
```

Reply:
As another point, you are encoding a math-formula, and using single-letter variable names in this context clarifies that this is a mathematical calculation, and that the types of those single-letter variables are mathematical objects like scalars and vectors. It is worthwhile to convey this with standardard mathematical-variable naming.

Also on readability, let me ask, which is easier to parse:

```
x = (-b + sqrt(b*b - 4*a*c)) / (2*a)
```

or:

```
x = (-linear_coefficient + sqrt(linear_coefficient * linear_coefficient - 
    4 * quadratic_coefficient * constant_term)) / (2 * quadratic_term)
```